engine/src/physics/solver/penetration_constraint/plugin.cpp

```diff
@@ -68,47 +77,90 @@ void solvePenetrationConstraint(Query<Entity, const Mass&, AccumulatedCorrection
             }
 
             // Relative velocity at contact
-            glm::vec3 vr2;
-            glm::vec3 vr1;
-            if (ent1 == constraint.entity)
-            {
-                vr2 = velocity2.vec;
-                vr1 = velocity1.vec;
-            }
-            else
+            glm::vec3 vr = velocity2.vec - velocity1.vec;
+
+            if (ent1 != constraint.entity)
             {
-                vr2 = velocity1.vec;
-                vr1 = velocity2.vec;
+                vr *= -1.0F;
             }
 
-            float vn = glm::dot(vr2 - vr1, constraint.normal);
+            float vn = glm::dot(vr, constraint.normal);
 
             // Compute normal impulse
             float impulse = -constraint.normalMass * massScale * (vn + bias) - impulseScale * totalImpulse;
 
             // Clamp the accumulated impulse
             float newImpulse = glm::max(totalImpulse + impulse, 0.0F);
             impulse = newImpulse - totalImpulse;
-            if (i == 0)
+            constraint.normalImpulse = newImpulse;
+
+            glm::vec3 p = impulse * constraint.normal;
+
+            if (ent1 != constraint.entity)
             {
-                correction1.impulse = newImpulse;
+                p *= -1.0F;
             }
-            else
+
+            v1 = velocity1.vec - p * mass1.inverseMass;
+            v2 = velocity2.vec + p * mass2.inverseMass;
+        }
+
+        // Friction
+        // for each contact point in the future, since we have particles now we only need to do it once
+        {
+            float normalImpulse = constraint.normalImpulse;
+            float frictionImpulse1 = constraint.frictionImpulse1;
+            float frictionImpulse2 = constraint.frictionImpulse2;
+
+            // Relative velocity at contact
+            glm::vec3 vr = velocity2.vec - velocity1.vec;
+
+            if (ent1 != constraint.entity)
             {
-                correction2.impulse = newImpulse;
+                vr *= -1.0F;
             }
 
-            glm::vec3 p = constraint.normal * impulse;
-            if (ent1 == constraint.entity)
+            glm::vec3 tangent1 = vr - glm::dot(vr, constraint.normal) * constraint.normal;
+            glm::vec3 tangent2;
+            float tangentLenSq = glm::length2(tangent1);
+            if (tangentLenSq > 1e-6 * 1e-6)
```

Comment:
I'm still a bit sketched out by these epsilon values, they're just guesses. But now the impact of incorrectly detecting it as 0 is much smaller, so lgtm